File: backend/app/api/routers/zadania.py
from typing import Annotated, Optional, Dict, Any, List

from fastapi import APIRouter, Depends, HTTPException, Body
from fastapi.responses import FileResponse
from starlette import status
from app.dependencies import get_current_user_from_cookie, get_zadania_service, get_pdf_service, any_logged_in_user, \
    kierownik_only, get_protokoly_service
from app.domain.requestsDTO import ZadanieUpdateDTO, ProtokolPodpisDTO
from app.models.models import Uzytkownik
from app.services.protokoly_service import ProtokolyService
from app.errors import SaveError
from app.schemas.user import User
from app.services.PDF_service import PDFService
from app.services.pdf_zadanie_service import render_zadanie_pdf  # Używamy tylko aliasu
from app.core.paths import PDF_DIR
from app.services.zadania_service import ZadaniaService

router = APIRouter(prefix="/api/zadania", tags=["Zadania"])



CurrentUserDep = Annotated[User, Depends(get_current_user_from_cookie)]


# ============= LISTA / SZCZEGÓŁY =============

@router.get("")
def list_zadania(
        zadania_service: ZadaniaService = Depends(get_zadania_service),
        dateFrom: str | None = None,
        dateTo: str | None = None,
        onlyOpen: bool = False,
        search: str | None = None,
        page: int = 1,
        page_size: int = 25,
        user: Uzytkownik = Depends(any_logged_in_user)
) -> Dict[str, Any]:
    """
    Zwraca listę zadań z paginacją i wyszukiwaniem.
    Korzysta z ZadaniaService, który deleguje wywołanie do Repo.

    Query params:
    - dateFrom: filtruj od daty
    - dateTo: filtruj do daty
    - onlyOpen: tylko otwarte zadania
    - search: wyszukaj po ID lub nazwie klienta
    - page: numer strony (domyślnie 1)
    - page_size: liczba elementów na stronę (domyślnie 25)

    Returns:
    - items: lista zadań na bieżącej stronie
    - total: całkowita liczba zadań (po filtrowaniu)
    - page: bieżąca strona
    - page_size: rozmiar strony
    - total_pages: całkowita liczba stron
    """
    return zadania_service.get_lista(
        date_from=dateFrom,
        date_to=dateTo,
        only_open=onlyOpen,
        search=search,
        page=page,
        page_size=page_size
    )

@router.patch("/patch/{znag_id}")
def patch_zadania(
        znag_id: int,
        update_dto: ZadanieUpdateDTO,
        service: ZadaniaService = Depends(get_zadania_service),
        user: Uzytkownik = Depends(any_logged_in_user)
):
    try:
        updated_zadanie = service.patch_zadanie(znag_id, update_dto)
        return updated_zadanie
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except RuntimeError as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

@router.post("/{znag_id}/podpis-wszystkie-protokoly")
def podpisz_wszystkie_protokoly(
        znag_id: int,
        podpis_dto: ProtokolPodpisDTO,
        protokoly_service: ProtokolyService = Depends(get_protokoly_service),
        user: Uzytkownik = Depends(any_logged_in_user)
):
    """Podpisuje wszystkie protokoły powiązane z danym zadaniem."""
    try:
        result = protokoly_service.zapisz_podpis_dla_wszystkich_protokolow_zadania(
            znag_id,
            podpis_dto.Podpis,
            podpis_dto.Klient
        )
        return result
    except SaveError as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.get("/{znag_id}")
def get_zadanie(
        znag_id: int,
        zadania_service: ZadaniaService = Depends(get_zadania_service),
        user: Uzytkownik = Depends(any_logged_in_user)
) -> Dict[str, Any]:
    """
    Zwraca pełne dane zadania (wszystkie kolumny włącznie z godzinami).
    """
    nag = zadania_service.get_naglowek_pelny(znag_id)
    if not nag:
        raise HTTPException(404, "Zadanie nie istnieje")
    return nag


@router.get("/{znag_id}/pozycje")
def get_zadanie_pozycje(
        current_user: CurrentUserDep,
        znag_id: int,
        zadania_service: ZadaniaService = Depends(get_zadania_service),
        user: Uzytkownik = Depends(any_logged_in_user)
    ):
    """
    Zwraca pozycje zadania (np. z widoku v_ZadaniePozycje). 
    Używa logiki z Serwisu, aby zastosować filtry zależne od roli użytkownika.
    """
    result = zadania_service.get_pozycje_by_user_role(current_user, znag_id)

    if not result:
        # Możesz chcieć zwrócić pustą listę zamiast 404, ale zachowałem logikę:
        raise HTTPException(status_code=404, detail="Pozycje nie istnieją")

    return result


# ============= ZARZĄDZANIE POZYCJĄ (flaga do przeglądu) =============

@router.put("/pozycje/{zpoz_id}/do-przegladu")
def set_do_przegladu(
        zpoz_id: int,
        zadania_service: ZadaniaService = Depends(get_zadania_service),
        payload: Dict[str, Any] = Body(...),
        user: Uzytkownik = Depends(kierownik_only)
):
    """
    Ustawia flagę 'do przeglądu' w wybranej pozycji zadania.
    """
    value = bool(payload.get("value"))
    user = payload.get("user")

    result = zadania_service.set_do_przegladu(zpoz_id, value, user)

    if not result:
        raise HTTPException(status_code=500, detail="Operacja nie powiodła się (rollback w DB)")

    return {"ok": True}


# ============= GENEROWANIE PDF =============

@router.post("/{znag_id}/pdf/generuj")
def generuj_pdf(
        znag_id: int,
        pdf_service: PDFService = Depends(get_pdf_service),
        body: dict | None = Body(None),
        user: Uzytkownik = Depends(any_logged_in_user)
):
    serwisanci = (body or {}).get("serwisanci") or []
    try:
        out_path = pdf_service.generuj_pdf_zadania(
            znag_id=znag_id,
            serwisanci=serwisanci
        )
    except Exception as e:
        # Obsługa innych błędów generowania
        raise HTTPException(500, detail="Błąd podczas generowania pliku PDF") from e

    return FileResponse(
        str(out_path),
        media_type="application/pdf",
        filename=f"zadanie_{znag_id}.pdf"
    )



@router.post("/{znag_id}/pdf/generuj/old")
def generuj_pdf_old(
        znag_id: int,
        zadania_service: ZadaniaService = Depends(get_zadania_service),
        body: dict | None = Body(None),
        user: Uzytkownik = Depends(any_logged_in_user)
):
    """
    Generuje i zwraca plik PDF dla danego zadania.
    """
    nagl = zadania_service.get_naglowek_by_id(znag_id)
    podpis = zadania_service.get_podpis(znag_id)
    if not nagl:
        raise HTTPException(404, "Zadanie nie istnieje")

    poz = zadania_service.get_pozycje_orm(znag_id)

    serwisanci = (body or {}).get("serwisanci") or []
    out_path = PDF_DIR / f"zadanie_{znag_id}.pdf"

    render_zadanie_pdf(
        out_path=str(out_path),
        naglowek=nagl,
        podpis=podpis,
        pozycje=poz,
        serwisanci=serwisanci
    )

    return FileResponse(
        str(out_path),
        media_type="application/pdf",
        filename=f"zadanie_{znag_id}.pdf"
    )

# ============= OPISY PRAC (dla awarii i prac różnych) =============

@router.get("/{znag_id}/opisy-prac")
def get_opisy_prac(
        znag_id: int,
        zadania_service: ZadaniaService = Depends(get_zadania_service),
        user: Uzytkownik = Depends(any_logged_in_user)
) -> List[Dict[str, Any]]:
    """
    Zwraca listę opisów prac dla zadania (awaria/prace różne).
    """
    opisy = zadania_service.get_opis_prac(znag_id)
    return [{"ZOP_Id": o.ZOP_Id, "ZOP_ZNAGL_Id": o.ZOP_ZNAGL_Id, "ZOP_OpisPrac": o.ZOP_OpisPrac} for o in opisy]


@router.post("/{znag_id}/opisy-prac")
def add_opis_prac(
        znag_id: int,
        payload: Dict[str, str] = Body(...),
        zadania_service: ZadaniaService = Depends(get_zadania_service),
        user: Uzytkownik = Depends(any_logged_in_user)
):
    """
    Dodaje nowy opis prac do zadania.
    """
    opis_prac = payload.get("opis_prac", "").strip()
    if not opis_prac:
        raise HTTPException(400, "Opis prac nie może być pusty")

    nowy = zadania_service.add_opis_prac(znag_id, opis_prac)
    return {"ZOP_Id": nowy.ZOP_Id, "ZOP_ZNAGL_Id": nowy.ZOP_ZNAGL_Id, "ZOP_OpisPrac": nowy.ZOP_OpisPrac}


@router.patch("/opisy-prac/{zop_id}")
def update_opis_prac(
        zop_id: int,
        payload: Dict[str, str] = Body(...),
        zadania_service: ZadaniaService = Depends(get_zadania_service),
        user: Uzytkownik = Depends(any_logged_in_user)
):
    """
    Aktualizuje opis prac.
    """
    opis_prac = payload.get("opis_prac", "").strip()
    if not opis_prac:
        raise HTTPException(400, "Opis prac nie może być pusty")

    updated = zadania_service.update_opis_prac(zop_id, opis_prac)
    if not updated:
        raise HTTPException(404, "Opis prac nie został znaleziony")

    return {"ZOP_Id": updated.ZOP_Id, "ZOP_ZNAGL_Id": updated.ZOP_ZNAGL_Id, "ZOP_OpisPrac": updated.ZOP_OpisPrac}


@router.delete("/opisy-prac/{zop_id}")
def delete_opis_prac(
        zop_id: int,
        zadania_service: ZadaniaService = Depends(get_zadania_service),
        user: Uzytkownik = Depends(any_logged_in_user)
):
    """
    Usuwa opis prac.
    """
    success = zadania_service.delete_opis_prac(zop_id)
    if not success:
        raise HTTPException(404, "Opis prac nie został znaleziony")

    return {"ok": True}

File: frontend/src/api/zadania.ts
// frontend/src/api/zadania.ts
import { api } from "./client";
import { Zadanie, ZadaniePozycja } from "../types";

/** Lista zadań (widok v_Zadania) z paginacją i wyszukiwaniem */
export async function getZadania(params?: {
  search?: string;
  page?: number;
  page_size?: number;
}): Promise<{
  items: Zadanie[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}> {
  const { data } = await api.get("/zadania", { params });
  return data;
}

/** (Opcjonalnie) pojedyncze zadanie – jeśli masz endpoint /api/zadania/:id */
export async function getZadanie(znagId: number): Promise<Zadanie> {
  const { data } = await api.get<Zadanie>(`/zadania/${znagId}`);
  return data;
}

/** Pozycje zadania (widok v_ZadaniePozycje) */
export async function getZadaniePozycje(
  znagId: number
): Promise<ZadaniePozycja[]> {
  const { data } = await api.get<ZadaniePozycja[]>(`/zadania/${znagId}/pozycje`);
  return data;
}

export async function getZadaniePozycjeSerwisant(
  znagId: number
): Promise<ZadaniePozycja[]> {
  const { data } = await api.get<ZadaniePozycja[]>(`/zadania/${znagId}/pozycje-serwisant`);
  return data;
}

/** Ustaw/wyłącz flagę 'do przeglądu' dla pozycji zadania */
export async function setDoPrzegladu(
  zpozId: number,
  value: boolean,
  user: string
): Promise<void> {
  await api.put(`/zadania/pozycje/${zpozId}/do-przegladu`, { value, user });
}

/** Generuj PDF dla zadania i zwróć Blob (nie zapisuje automatycznie) */
export async function generateZadaniePdf(
  znagId: number,
  serwisanci: string[] = []
): Promise<Blob> {
  const { data } = await api.post(
    `/zadania/${znagId}/pdf/generuj`, // <-- BYŁO: /pdf
    { serwisanci },
    { responseType: "blob" }
  );
  return data as Blob;
}

/** Generuj i od razu pobierz PDF (helper do przycisku w UI) */
export async function downloadZadaniePdf(
  znagId: number,
  serwisanci: string[] = []
): Promise<void> {
  const blob = await generateZadaniePdf(znagId, serwisanci);
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = `zadanie_${znagId}.pdf`;
  document.body.appendChild(a);
  a.click();
  a.remove();
  URL.revokeObjectURL(url);
}


export async function patchZadanie(
  znagId: number,
  fieldName: string,
  fieldValue: string
): Promise<Zadanie> {
  const { data } = await api.patch(
    `/zadania/patch/${znagId}`,
    {[fieldName]: fieldValue}
  );
  return data;
}

export async function podpiszZadanie(
  znagId: number,
  podpisKlienta: string
): Promise<Zadanie>{
  const { data } = await api.patch(
    `/zadania/patch/${znagId}`,
    {ZNAG_KlientPodpis: podpisKlienta}
  )
  return data;
}

export async function patchZadanieMultiple(
  znagId: number,
  fields: Record<string, any>
): Promise<Zadanie> {
  const { data } = await api.patch(
    `/zadania/patch/${znagId}`,
    fields
  );
  return data;
}

export async function podpiszWszystkieProtokoly(
  znagId: number,
  podpisKlienta: string,
  klient: string = "Klient"
): Promise<{ ok: boolean; signed_count: number }> {
  const { data } = await api.post(
    `/zadania/${znagId}/podpis-wszystkie-protokoly`,
    { Podpis: podpisKlienta, Klient: klient }
  );
  return data;
}

export async function getProtokolyZadania(
  znagId: number
): Promise<any[]> {
  const { data } = await api.get(`/zadania/${znagId}/protokoly`);
  return data;
}

// ============= OPISY PRAC =============

export interface OpisPrac {
  ZOP_Id: number;
  ZOP_ZNAGL_Id: number;
  ZOP_OpisPrac: string;
}

export async function getOpisyPrac(znagId: number): Promise<OpisPrac[]> {
  const { data } = await api.get(`/zadania/${znagId}/opisy-prac`);
  return data;
}

export async function addOpisPrac(znagId: number, opisPrac: string): Promise<OpisPrac> {
  const { data } = await api.post(`/zadania/${znagId}/opisy-prac`, { opis_prac: opisPrac });
  return data;
}

export async function updateOpisPrac(zopId: number, opisPrac: string): Promise<OpisPrac> {
  const { data } = await api.patch(`/zadania/opisy-prac/${zopId}`, { opis_prac: opisPrac });
  return data;
}

export async function deleteOpisPrac(zopId: number): Promise<void> {
  await api.delete(`/zadania/opisy-prac/${zopId}`);
}